File: api/save.py
import os
import json
from http.server import BaseHTTPRequestHandler
import urllib.request
import urllib.error

class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        try:
            # Read request body
            content_length = int(self.headers.get('Content-Length', 0))
            body = self.rfile.read(content_length)
            data = json.loads(body)
            
            block_id = data.get('block_id')
            content = data.get('content')
            
            if not block_id or content is None:
                self._send_response(400, {'error': 'Missing block_id or content'})
                return
            
            # Get token from environment
            token = os.environ.get('GITHUB_TOKEN')
            if not token:
                self._send_response(500, {'error': 'Server misconfiguration: No token'})
                return
            
            # Dispatch to GitHub Actions
            owner = os.environ.get('REPO_OWNER', 'EliezerRosa')
            repo = os.environ.get('REPO_NAME', 'RVM-Designacoes-Antigravity')
            
            dispatch_url = f'https://api.github.com/repos/{owner}/{repo}/dispatches'
            
            payload = json.dumps({
                'event_type': 'atomic_write',
                'client_payload': {
                    'block_id': block_id,
                    'content': content
                }
            }).encode('utf-8')
            
            req = urllib.request.Request(
                dispatch_url,
                data=payload,
                headers={
                    'Authorization': f'token {token}',
                    'Accept': 'application/vnd.github.v3+json',
                    'Content-Type': 'application/json',
                    'User-Agent': 'RVM-Gateway'
                },
                method='POST'
            )
            
            try:
                with urllib.request.urlopen(req, timeout=10) as response:
                    if response.status == 204:
                        self._send_response(200, {
                            'status': 'accepted',
                            'message': 'Transação enfileirada no Atomic Writer',
                            'block_id': block_id
                        })
                    else:
                        self._send_response(response.status, {'error': 'Unexpected response from GitHub'})
            except urllib.error.HTTPError as e:
                self._send_response(e.code, {'error': f'GitHub API error: {e.reason}'})
            except urllib.error.URLError as e:
                self._send_response(502, {'error': f'Failed to reach GitHub: {str(e)}'})
                
        except json.JSONDecodeError:
            self._send_response(400, {'error': 'Invalid JSON'})
        except Exception as e:
            self._send_response(500, {'error': str(e)})
    
    def do_OPTIONS(self):
        # Handle CORS preflight
        self.send_response(200)
        self._send_cors_headers()
        self.end_headers()
    
    def _send_response(self, status_code, data):
        self.send_response(status_code)
        self._send_cors_headers()
        self.send_header('Content-Type', 'application/json')
        self.end_headers()
        self.wfile.write(json.dumps(data).encode('utf-8'))
    
    def _send_cors_headers(self):
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')

File: api/save.js
// Vercel Serverless Function (Node.js runtime, not Edge)
// Path: /api/save

export default async function handler(req, res) {
    // Set CORS headers
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

    // Handle CORS preflight
    if (req.method === 'OPTIONS') {
        return res.status(200).end();
    }

    if (req.method !== 'POST') {
        return res.status(405).json({ error: 'Method not allowed' });
    }

    try {
        const { block_id, content } = req.body;

        if (!block_id || content === undefined) {
            return res.status(400).json({ error: 'Missing block_id or content' });
        }

        const token = process.env.GITHUB_TOKEN;
        if (!token) {
            return res.status(500).json({ error: 'Server misconfiguration: No token' });
        }

        const owner = process.env.REPO_OWNER || 'EliezerRosa';
        const repo = process.env.REPO_NAME || 'RVM-Designacoes-Antigravity';

        const githubResponse = await fetch(
            `https://api.github.com/repos/${owner}/${repo}/dispatches`,
            {
                method: 'POST',
                headers: {
                    'Authorization': `token ${token}`,
                    'Accept': 'application/vnd.github.v3+json',
                    'Content-Type': 'application/json',
                    'User-Agent': 'RVM-Gateway',
                },
                body: JSON.stringify({
                    event_type: 'atomic_write',
                    client_payload: { block_id, content },
                }),
            }
        );

        if (githubResponse.status === 204) {
            return res.status(200).json({
                status: 'accepted',
                message: 'Transação enfileirada no Atomic Writer',
                block_id,
            });
        } else {
            const errorText = await githubResponse.text();
            return res.status(githubResponse.status).json({ error: `GitHub API error: ${errorText}` });
        }
    } catch (e) {
        return res.status(500).json({ error: e.message });
    }
}
